File: benchmarks/bench_namespace_calls.pl
#!/usr/bin/perl
# Perl equivalent of bench_namespace_calls.stash
use strict;
use warnings;
use Time::HiRes qw(time);
use POSIX qw(floor);

my $ITERATIONS = 200000;

print "=== Benchmark: Namespace Calls ===\n";

my $start = time();
my $i = 0;
my $acc = 0.0;
while ($i < $ITERATIONS) {
    # math namespace
    my $sq = sqrt($i + 1);
    my $ab = abs(-$i - 1);
    my $fl = floor($sq + 0.5);
    my $pw = 2 ** 8;
    my $mx = $i > $i + 1 ? $i : $i + 1;
    my $mn = $i < $i + 1 ? $i : $i + 1;

    # str namespace
    my $up = uc("hello");
    my $lo = lc("WORLD");
    my $tr = "  stash  " =~ s/^\s+|\s+$//gr;
    my $rp = "abc" =~ s/b/B/gr;

    # conv namespace
    my $cs  = "$i";
    my $ci  = int("42");
    my $chx = sprintf("%x", 255);

    $acc = $acc + $sq + $ab + $fl + $pw + $mx + $mn;
    $i++;
}
my $elapsed = int((time() - $start) * 1000);

print "Iterations: $ITERATIONS\n";
print "Result (checksum): $acc\n";
print "Time: $elapsed ms\n";

File: benchmarks/bench_namespace_calls.js
// Node.js equivalent of bench_namespace_calls.stash
// =============================================================================
// Benchmark: Namespace Member Lookup
// Tests the cost of resolving built-in functions across different namespaces
// (math.*, str.*, conv.*). Each iteration performs multiple lookups to stress
// the dispatch path.
// =============================================================================

const ITERATIONS = 200000;

console.log("=== Benchmark: Namespace Calls ===");

let start = Date.now();
let i = 0;
let acc = 0.0;
while (i < ITERATIONS) {
    // math namespace
    let sq  = Math.sqrt(i + 1);
    let ab  = Math.abs(-i - 1);
    let fl  = Math.floor(sq + 0.5);
    let pw  = Math.pow(2, 8);
    let mx  = Math.max(i, i + 1);
    let mn  = Math.min(i, i + 1);

    // str namespace
    let up  = "hello".toUpperCase();
    let lo  = "WORLD".toLowerCase();
    let tr  = "  stash  ".trim();
    let rp  = "abc".replace("b", "B");

    // conv namespace
    let cs  = String(i);
    let ci  = parseInt("42");
    let chx = (255).toString(16);

    acc = acc + sq + ab + fl + pw + mx + mn;
    i++;
}
let elapsed = Date.now() - start;

console.log("Iterations: " + ITERATIONS);
console.log("Result (checksum): " + acc);
console.log("Time: " + elapsed + " ms");
